package/MDAnalysis/coordinates/RDKit.py

```diff
@@ -95,10 +133,538 @@ def __init__(self, filename, **kwargs):
+            raise TypeError("No `atoms` attribute in object of type {}, "
+                            "please use a valid AtomGroup or Universe".format(
+                                type(obj))) from None
+
+        try:
+            elements = ag.elements
+        except NoDataError:
+            raise AttributeError(
+                "The `elements` attribute is required for the RDKitConverter "
+                "but is not present in this AtomGroup. Please refer to the "
+                "documentation to guess elements from other attributes or "
+                "type `help(mda.topology.guessers)`") from None
+
+        if "H" not in ag.elements:
+            warnings.warn(
+                "No hydrogen atom could be found in the topology, but the "
+                "converter requires all hydrogens to be explicit. Please "
+                "check carefully the output molecule as the converter is "
+                "likely to add negative charges and assign incorrect bond "
+                "orders to structures with implicit hydrogens."
+            )
+
+        # attributes accepted in PDBResidueInfo object
+        pdb_attrs = {}
+        if hasattr(ag, "bfactors") and hasattr(ag, "tempfactors"):
+            raise AttributeError(
+                "Both `tempfactors` and `bfactors` attributes are present but "
+                "only one can be assigned to the RDKit molecule. Please "
+                "delete the unnecessary one and retry."
+            )
+        for attr in RDATTRIBUTES.keys():
+            if hasattr(ag, attr):
+                pdb_attrs[attr] = getattr(ag, attr)
+
+        other_attrs = {}
+        for attr in ["charges", "segids", "types"]:
+            if hasattr(ag, attr):
+                other_attrs[attr] = getattr(ag, attr)
+
+        mol = Chem.RWMol()
+        # map index in universe to index in mol
+        atom_mapper = {}
+
+        for i, (atom, element) in enumerate(zip(ag, elements)):
+            # create atom
+            rdatom = Chem.Atom(element.capitalize())
+            # disable adding H to the molecule
+            rdatom.SetNoImplicit(NoImplicit)
+            # add PDB-like properties
+            mi = Chem.AtomPDBResidueInfo()
+            for attr, values in pdb_attrs.items():
+                _add_mda_attr_to_rdkit(attr, values[i], mi)
+            rdatom.SetMonomerInfo(mi)
+            # other properties
+            for attr in other_attrs.keys():
+                value = other_attrs[attr][i]
+                attr = "_MDAnalysis_%s" % _TOPOLOGY_ATTRS[attr].singular
+                _set_atom_property(rdatom, attr, value)
+            _set_atom_property(rdatom, "_MDAnalysis_index", int(atom.ix))
+            # add atom
+            index = mol.AddAtom(rdatom)
+            atom_mapper[atom.ix] = index
+
+        try:
+            if (len(ag.bonds) == 0) and (ag.n_atoms > 1):
+                # force guessing bonds
+                raise NoDataError
+        except NoDataError:
```

Comment:
What is the motivation for raising `NoDataError` and then capturing it right away to issue a warning and guess the bonds vs. just directly issuing the warning and guessing the bonds when the conditional is satisfied?

Reply:
I do 2 things here: I implicitly check that ag.bonds exist, if it doesn't it will raise the NoDataError. And sometimes ag.bonds exist but is empty (PDB file without a CONECT record) so I need the if statement as well, and to avoid duplicate code I raise the same NoDataError. All this was prior to #2855 so I can safely remove the if statement now

Reply:
I see, yeah that probably deserved a comment at least, since it is a bit confusing as written. The repetition might have even been 'ok' to avoid the confusion (i.e., clarity over concision).

Sounds like you can simplify this now IIRC.